File: apps/web/src/lib/api/catalog-categories.ts
import type { Category } from "@/lib/mock-data/categories";

/** Raw category shape from FastAPI `/api/v1/catalog/categories`. */
export type ApiCatalogCategory = {
  id: string;
  company_id: string;
  parent_id: string | null;
  path: string;
  depth: number;
  name: string;
  caption: string;
  slug: string;
  sort_order: number;
  product_count: number;
  is_active: boolean;
  show_in_top_menu: boolean;
  description: string | null;
  meta_title: string | null;
  meta_description: string | null;
  meta_keywords: string | null;
  icon_url: string | null;
  banner_url: string | null;
  icon_media_id: string | null;
  banner_media_id: string | null;
  created_at: string;
  updated_at: string;
};

export type ApiCategoryListResponse = {
  data: ApiCatalogCategory[];
  meta: { count: number };
  errors: string[];
};

export type ApiCategoryResponse = {
  data: ApiCatalogCategory;
  errors: string[];
};

export function apiCategoryToCategory(row: ApiCatalogCategory): Category {
  return {
    id: row.id,
    name: row.name,
    caption: row.caption || row.name,
    slug: row.slug,
    parentId: row.parent_id,
    sortOrder: row.sort_order,
    productCount: row.product_count,
    active: row.is_active,
    showInTopMenu: row.show_in_top_menu,
    updatedAt: row.updated_at.slice(0, 10),
    description: row.description ?? undefined,
    metaTitle: row.meta_title ?? undefined,
    metaDescription: row.meta_description ?? undefined,
    metaKeywords: row.meta_keywords ?? undefined,
    iconUrl: row.icon_url ?? undefined,
    bannerUrl: row.banner_url ?? undefined,
    iconMediaId: row.icon_media_id ?? undefined,
    bannerMediaId: row.banner_media_id ?? undefined,
  };
}

export type CreateCatalogCategoryInput = {
  name: string;
  caption?: string;
  slug: string;
  parentId?: string | null;
  active?: boolean;
  showInTopMenu?: boolean;
  description?: string;
  metaTitle?: string;
  metaDescription?: string;
  metaKeywords?: string;
  iconUrl?: string;
  bannerUrl?: string;
  iconMediaId?: string;
  bannerMediaId?: string;
};

export type UpdateCatalogCategoryInput = Partial<CreateCatalogCategoryInput>;

function categoryInputToApiPayload(input: CreateCatalogCategoryInput | UpdateCatalogCategoryInput) {
  const payload: Record<string, unknown> = {};
  if ("name" in input && input.name !== undefined) payload.name = input.name;
  if ("caption" in input && input.caption !== undefined) payload.caption = input.caption;
  if ("slug" in input && input.slug !== undefined) payload.slug = input.slug;
  if ("parentId" in input && input.parentId !== undefined) payload.parent_id = input.parentId;
  if ("active" in input && input.active !== undefined) payload.is_active = input.active;
  if ("showInTopMenu" in input && input.showInTopMenu !== undefined) {
    payload.show_in_top_menu = input.showInTopMenu;
  }
  if ("description" in input && input.description !== undefined) {
    payload.description = input.description;
  }
  if ("metaTitle" in input && input.metaTitle !== undefined) payload.meta_title = input.metaTitle;
  if ("metaDescription" in input && input.metaDescription !== undefined) {
    payload.meta_description = input.metaDescription;
  }
  if ("metaKeywords" in input && input.metaKeywords !== undefined) {
    payload.meta_keywords = input.metaKeywords;
  }
  if ("iconUrl" in input && input.iconUrl !== undefined) payload.icon_url = input.iconUrl;
  if ("bannerUrl" in input && input.bannerUrl !== undefined) payload.banner_url = input.bannerUrl;
  if ("iconMediaId" in input && input.iconMediaId !== undefined) {
    payload.icon_media_id = input.iconMediaId;
  }
  if ("bannerMediaId" in input && input.bannerMediaId !== undefined) {
    payload.banner_media_id = input.bannerMediaId;
  }
  return payload;
}

export function categoryToApiCreatePayload(input: CreateCatalogCategoryInput) {
  return categoryInputToApiPayload(input);
}

export function categoryToApiUpdatePayload(input: UpdateCatalogCategoryInput) {
  return categoryInputToApiPayload(input);
}

export function partialCategoryToApiUpdate(data: Partial<Category>): UpdateCatalogCategoryInput {
  const input: UpdateCatalogCategoryInput = {};
  if (data.name !== undefined) input.name = data.name;
  if (data.caption !== undefined) input.caption = data.caption;
  if (data.slug !== undefined) input.slug = data.slug;
  if (data.parentId !== undefined) input.parentId = data.parentId;
  if (data.active !== undefined) input.active = data.active;
  if (data.showInTopMenu !== undefined) input.showInTopMenu = data.showInTopMenu;
  if (data.description !== undefined) input.description = data.description;
  if (data.metaTitle !== undefined) input.metaTitle = data.metaTitle;
  if (data.metaDescription !== undefined) input.metaDescription = data.metaDescription;
  if (data.metaKeywords !== undefined) input.metaKeywords = data.metaKeywords;
  if (data.iconUrl !== undefined) input.iconUrl = data.iconUrl;
  if (data.bannerUrl !== undefined) input.bannerUrl = data.bannerUrl;
  if (data.iconMediaId !== undefined) input.iconMediaId = data.iconMediaId;
  if (data.bannerMediaId !== undefined) input.bannerMediaId = data.bannerMediaId;
  return input;
}

File: apps/api/app/routers/catalog_categories.py
from __future__ import annotations

from typing import Dict, List, Optional, Set

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.deps.auth import require_write_access
from app.database import get_db
from app.models.catalog_category import CatalogCategory
from app.models.catalog_product import CatalogProduct
from app.schemas.category import (
    CategoryCreate,
    CategoryListMeta,
    CategoryListResponse,
    CategoryRead,
    CategoryReorderRequest,
    CategoryResponse,
    CategoryTreeNode,
    CategoryTreeResponse,
    CategoryUpdate,
)

router = APIRouter(prefix="/categories", tags=["catalog-categories"])


def _product_counts_by_name(db: Session) -> Dict[str, int]:
    rows = (
        db.query(CatalogProduct.category, CatalogProduct.id)
        .filter(CatalogProduct.category.isnot(None))
        .all()
    )
    counts: Dict[str, int] = {}
    for name, _ in rows:
        if name:
            counts[name] = counts.get(name, 0) + 1
    return counts


def _to_read(row: CatalogCategory, product_counts: Dict[str, int]) -> CategoryRead:
    data = CategoryRead.model_validate(row)
    data.product_count = product_counts.get(row.name, 0)
    return data


def _validate_parent(
    db: Session,
    category_id: Optional[str],
    parent_id: Optional[str],
) -> Optional[CatalogCategory]:
    if parent_id is None:
        return None

    if category_id and parent_id == category_id:
        raise HTTPException(status_code=400, detail="Category cannot be its own parent")

    parent = db.get(CatalogCategory, parent_id)
    if not parent:
        raise HTTPException(status_code=404, detail="Parent category not found")

    if category_id:
        current = db.get(CatalogCategory, category_id)
        if current and current.path.startswith(f"{parent.path}{category_id}/"):
            raise HTTPException(status_code=400, detail="Cannot move category under its descendant")

    return parent


def _apply_path_depth(row: CatalogCategory, parent: Optional[CatalogCategory]) -> None:
    if parent:
        row.parent_id = parent.id
        row.depth = parent.depth + 1
        row.path = f"{parent.path}{row.id}/"
    else:
        row.parent_id = None
        row.depth = 0
        row.path = f"/{row.id}/"


def _refresh_descendant_paths(db: Session, row: CatalogCategory) -> None:
    children = db.query(CatalogCategory).filter(CatalogCategory.parent_id == row.id).all()
    for child in children:
        child.depth = row.depth + 1
        child.path = f"{row.path}{child.id}/"
        _refresh_descendant_paths(db, child)


def _next_sort_order(db: Session, parent_id: Optional[str]) -> int:
    query = db.query(CatalogCategory.sort_order).filter(
        CatalogCategory.parent_id == parent_id,
    )
    max_order = query.order_by(CatalogCategory.sort_order.desc()).first()
    return (max_order[0] + 1) if max_order else 0


def _collect_descendant_ids(db: Session, root_ids: List[str]) -> Set[str]:
    to_delete: Set[str] = set()
    queue = list(root_ids)
    while queue:
        current = queue.pop()
        if current in to_delete:
            continue
        to_delete.add(current)
        children = (
            db.query(CatalogCategory.id)
            .filter(CatalogCategory.parent_id == current)
            .all()
        )
        queue.extend(child[0] for child in children)
    return to_delete


def _build_tree(rows: List[CategoryRead]) -> List[CategoryTreeNode]:
    nodes: Dict[str, CategoryTreeNode] = {}
    for row in rows:
        nodes[row.id] = CategoryTreeNode(**row.model_dump(), children=[])

    roots: List[CategoryTreeNode] = []
    for node in nodes.values():
        if node.parent_id and node.parent_id in nodes:
            nodes[node.parent_id].children.append(node)
        else:
            roots.append(node)

    def sort_children(node: CategoryTreeNode) -> None:
        node.children.sort(key=lambda c: c.sort_order)
        for child in node.children:
            sort_children(child)

    roots.sort(key=lambda c: c.sort_order)
    for root in roots:
        sort_children(root)
    return roots


@router.get("", response_model=CategoryListResponse)
def list_categories(db: Session = Depends(get_db)) -> CategoryListResponse:
    rows = db.query(CatalogCategory).order_by(CatalogCategory.path, CatalogCategory.sort_order).all()
    product_counts = _product_counts_by_name(db)
    data = [_to_read(r, product_counts) for r in rows]
    return CategoryListResponse(data=data, meta=CategoryListMeta(count=len(data)))


@router.get("/tree", response_model=CategoryTreeResponse)
def category_tree(db: Session = Depends(get_db)) -> CategoryTreeResponse:
    rows = db.query(CatalogCategory).order_by(CatalogCategory.path, CatalogCategory.sort_order).all()
    product_counts = _product_counts_by_name(db)
    flat = [_to_read(r, product_counts) for r in rows]
    return CategoryTreeResponse(data=_build_tree(flat))


@router.patch("/reorder", response_model=CategoryListResponse)
def reorder_categories(
    payload: CategoryReorderRequest,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> CategoryListResponse:
    rows = (
        db.query(CatalogCategory)
        .filter(CatalogCategory.parent_id == payload.parent_id)
        .all()
    )
    row_map = {r.id: r for r in rows}

    if set(payload.ordered_ids) != set(row_map.keys()):
        raise HTTPException(status_code=400, detail="ordered_ids must match sibling categories")

    for index, category_id in enumerate(payload.ordered_ids):
        row_map[category_id].sort_order = index

    db.commit()

    all_rows = db.query(CatalogCategory).order_by(CatalogCategory.path, CatalogCategory.sort_order).all()
    product_counts = _product_counts_by_name(db)
    data = [_to_read(r, product_counts) for r in all_rows]
    return CategoryListResponse(data=data, meta=CategoryListMeta(count=len(data)))


@router.get("/{category_id}", response_model=CategoryResponse)
def get_category(category_id: str, db: Session = Depends(get_db)) -> CategoryResponse:
    row = db.get(CatalogCategory, category_id)
    if not row:
        raise HTTPException(status_code=404, detail="Category not found")
    product_counts = _product_counts_by_name(db)
    return CategoryResponse(data=_to_read(row, product_counts))


@router.post("", response_model=CategoryResponse, status_code=201)
def create_category(
    payload: CategoryCreate,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> CategoryResponse:
    existing = db.query(CatalogCategory).filter(CatalogCategory.slug == payload.slug).first()
    if existing:
        raise HTTPException(status_code=409, detail="Category slug already exists")

    parent = _validate_parent(db, None, payload.parent_id)
    row = CatalogCategory(
        name=payload.name,
        caption=payload.caption or payload.name,
        slug=payload.slug,
        sort_order=payload.sort_order,
        is_active=payload.is_active,
        show_in_top_menu=payload.show_in_top_menu,
        description=payload.description,
        meta_title=payload.meta_title,
        meta_description=payload.meta_description,
        meta_keywords=payload.meta_keywords,
        icon_url=payload.icon_url,
        banner_url=payload.banner_url,
        icon_media_id=payload.icon_media_id,
        banner_media_id=payload.banner_media_id,
    )
    db.add(row)
    db.flush()

    if payload.sort_order == 0:
        row.sort_order = _next_sort_order(db, parent.id if parent else None)

    _apply_path_depth(row, parent)
    db.commit()
    db.refresh(row)

    product_counts = _product_counts_by_name(db)
    return CategoryResponse(data=_to_read(row, product_counts))


@router.patch("/{category_id}", response_model=CategoryResponse)
def update_category(
    category_id: str,
    payload: CategoryUpdate,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> CategoryResponse:
    row = db.get(CatalogCategory, category_id)
    if not row:
        raise HTTPException(status_code=404, detail="Category not found")

    changes = payload.model_dump(exclude_unset=True)
    if not changes:
        product_counts = _product_counts_by_name(db)
        return CategoryResponse(data=_to_read(row, product_counts))

    slug = changes.get("slug", row.slug)
    if slug != row.slug:
        conflict = (
            db.query(CatalogCategory)
            .filter(CatalogCategory.slug == slug, CatalogCategory.id != category_id)
            .first()
        )
        if conflict:
            raise HTTPException(status_code=409, detail="Category slug already exists")

    parent_changed = "parent_id" in changes and changes["parent_id"] != row.parent_id
    if parent_changed:
        parent = _validate_parent(db, category_id, changes["parent_id"])
        _apply_path_depth(row, parent)
        _refresh_descendant_paths(db, row)

    for key, value in changes.items():
        if key == "parent_id":
            continue
        setattr(row, key, value)

    db.commit()
    db.refresh(row)

    product_counts = _product_counts_by_name(db)
    return CategoryResponse(data=_to_read(row, product_counts))


@router.delete("/{category_id}", status_code=204)
def delete_category(
    category_id: str,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> None:
    row = db.get(CatalogCategory, category_id)
    if not row:
        raise HTTPException(status_code=404, detail="Category not found")

    ids = _collect_descendant_ids(db, [category_id])
    db.query(CatalogCategory).filter(CatalogCategory.id.in_(ids)).delete(synchronize_session=False)
    db.commit()
